Add `focus-metavariable:` operator

IagoAbal opened the pull request:
This operator allows to focus matching on a given metavariable, within a
`patterns:` operator.

Closes #4453

test plan:
make test # test included

PR checklist:

- [ ] Documentation is up-to-date
- [ ] Changelog is up-to-date
- [x] Change has no security implications (otherwise, ping security team)

IagoAbal committed:
Add `focus-metavariable:` operator

This operator allows to focus matching on a given metavariable, within a
`patterns:` operator.

Closes #4453

test plan:
make test # test included

aryx commented on semgrep-core/src/core/Rule.ml:
```diff
@@ -109,8 +109,13 @@ type formula =
-  (* see Match_rules.split_and() *)
-  | And of tok * formula list * (tok * metavar_cond) list
+  (* see Specialize_formula.split_and() *)
+  | And of
```
maybe we can introduce a record at this point.

ievans reviewed (commented):
@IagoAbal can we hold this for a bit and try to get some external user feedback on it? It is really not obvious to me what the operator does, even reading through the example, which concerns me as a long-time Semgrep user. Do you have a docs PR that I could share with some users?

Also I feel like a test for the JSON output of this would be good

IagoAbal commented:
> @IagoAbal can we hold this for a bit and try to get some external user feedback on it? It is really not obvious to me what the operator does, even reading through the example, which concerns me as a long-time Semgrep user. Do you have a docs PR that I could share with some users?

No docs PR yet, I will work on one. But given a `focus-metavariable: $X` operator, whatever your patterns match it will be intersected with the range of what `$X` matches. So you can "focus" the match on a specific metavariable.

ievans commented:
I think the main thing I'm missing is what's different from adding a `pattern: $X`? (just performance or are there other implications)

Isaac0616 commented:
Not sure about `focus-metavariable` but only adding `pattern: $X` doesn't quite work because, for example,
```
patterns:
   - pattern-inside: 'requests.get(..., verify=$X, ...)'
   - pattern: $X
```
will match every `False` in `request.get(url, verify=False, foo=False, bar=False)`.

aryx commented:
II guess we can add poning's examples in the docs, so the difference are clear.

IagoAbal commented:
Oh OK, I can add a simpler test case that exemplifies this in a more obvious way.

```
patterns:
    - pattern-inside: foo($X, ...)
    - focus-metavariable: $X
    - pattern: 1
```

matches

```
foo (1,  # here
     2)  # not here
```

but it does not match

```
foo (2,
     1)
```

also it would match:

```
foo (1,  # here
     1)  # not here
```

whereas `pattern: $X` would have matched all the occurrences of `1`.

IagoAbal committed:
Yoann's review

IagoAbal committed:
Isaac's review

Removed changelog entry for now, as we want to get some feedback from
selected users first.
